### PR TITLE
URL-encode/decode resource names for HTTP API

### PR DESCRIPTION
**Pull request by littlestar642**
### Overview

Relates to: #11258

To encode/decode the path params in HTTP API, we can use golang's `url.PathEscape` and `url.PathUnescape` methods.

url.PathEscape - PathEscape escapes the string so it can be safely placed inside a URL path segment, replacing special characters (including /) with %XX sequences as needed.

### CODE REVIEW

**Review by kisunji — APPROVED** [review]
LGTM! Thanks for this

**Comment by kisunji** [comment]
Will push a commit to add a changelog to this PR then commit

File: agent/agent_endpoint.go
Comment (jkirschner-hashicorp) [inline agent/agent_endpoint.go:367]: I recommend that we simplify this to make it easier to maintain and apply consistently to all endpoints (remove redundant operations and boilerplate). I think this can be solved by abstracting into a function call like:
```
func GetPathSuffixUnescaped(path string, prefixToTrim string) (string, error) {
  // The suffix may be URL-encoded, so attempt to decode
  suffixRaw := strings.TrimPrefix(path, prefixToTrim)
  suffixUnescaped, err := url.PathUnescape(suffixRaw)
  
  if err != nil {
    return suffixRaw, fmt.Errorf("failure in unescaping path param %q: %v", suffixRaw, err)
  }
  
  return suffixUnescaped, nil
}
```

Then, each endpoint replacement would look something like this:
```suggestion
	id, err := GetPathSuffixUnescaped(req.URL.Path, "/v1/agent/service/")
	if err != nil {
		s.agent.logger.Error(err)
	}
```

---

@kisunji : I'm not sure the best place in the source code to put a shared function like this. Thoughts on that? And whether you agree with this code change suggestion?

@littlestar642: are you interested in making such an enhancement to the PR? If not (we realize some time has passed since you posted this), let us know!

**Review by jkirschner-hashicorp — COMMENTED** [review]
> PR ready to merge, awaiting decision to backport single change or batch them in next release

@kisunji, @littlestar642: we decided we'll try to merge the changes just for this endpoint, then internally consider (manually) backporting to 1.9 and 1.10.

Since there's some work that goes into manual backporting, I think it makes sense to address the suggested change in my review first (as it should make rolling out to other endpoints easier). @littlestar642: let me know if you're interested in making that change (before we merge and then begin the backport process).

File: agent/agent_endpoint.go
Comment (kisunji) [reply agent/agent_endpoint.go:367]: Great suggestion! I think `agent/http.go` is a reasonable place for a general function that can be reused by all endpoints.

**Comment by jkirschner-hashicorp** [comment]
@kisunji : I rebased and made the change I proposed. This is ready for your final review.

File: agent/agent_endpoint.go
Comment (jkirschner-hashicorp) [reply agent/agent_endpoint.go]: Thanks - change made! I also added a unit test for that function.

File: agent/agent_endpoint.go
Comment (jkirschner-hashicorp) [inline agent/agent_endpoint.go:383]: Ran this against a consul binary just to see for myself that it worked.

```
# Successfully decoded!
$ curl --request GET localhost:8500/v1/agent/service/counting%2Fid
unknown service ID: counting/idj
```

File: agent/agent_endpoint.go
Comment (jkirschner-hashicorp) [inline agent/agent_endpoint.go:385]: When I ran against a consul binary with an invalid encoded value, this is what happened:
```
$ curl --request GET localhost:8500/v1/agent/service/counting%2id
400 Bad Request

# No error appears in consul log output
```

@kisunji : Is this what we'd expect/want to happen?

Comment (kisunji) [reply agent/agent_endpoint.go:385]: Summarizing our discussion over slack: most middlewares reject invalid URLs early so the request never reaches the handler. It's almost impossible to externally test the error check via an HTTP client because you either: 1) face early rejection by go's http lib; or 2) encode into a valid url and never hit the error code path.

File: agent/http.go
Comment (kisunji) [inline agent/http.go:1130]:
```suggestion
		return suffixRaw, fmt.Errorf("cannot unescape path param %q: %w", suffixRaw, err)
```

**Review by kisunji — CHANGES_REQUESTED** [review]
Looks good! @jkirschner-hashicorp We discussed that we should return early on error from `getPathSuffixUnescaped`. Once that change is made we can merge 👍

File: agent/http.go
Comment (jkirschner-hashicorp) [reply agent/http.go:1130]: @kisunji : I agree with the wording change, but not sure I agree with the change from `%v` to `%w` for the error.

Per [this Golang page](https://go.dev/blog/go1.13-errors):
> ...wrapping an error makes that error part of your API. If you don’t want to commit to supporting that error as part of your API in the future, you shouldn’t wrap the error.

> It’s important to remember that whether you wrap or not, the error text will be the same. A person trying to understand the error will have the same information either way; the choice to wrap is about whether to give programs additional information so they can make more informed decisions, or to withhold that information to preserve an abstraction layer.

In this case, I don't think the caller needs to be able to able to unwrap the returned error to specifically look for the unescape error.

Comment (kisunji) [reply agent/http.go:1130]: That's totally fair! In this case it doesn't matter too much since our http handler ultimately consumes all the errors before returning plaintext or json to the client but certainly a good thing to keep in mind.

File: agent/agent_endpoint.go
Comment (jkirschner-hashicorp) [inline agent/agent_endpoint.go:1330]: Why do we sometime use `return nil, err` within these functions, and other times use:
```
		fmt.Fprint(resp, "some error message")
		return nil, nil
```

?

Comment (jkirschner-hashicorp) [reply agent/agent_endpoint.go:1330]: @kisunji, @rboyer, @kyhavlov : any ideas why this is? (I want to make sure my use of `return nil, err` is what I should be doing in these cases.)

Comment (kisunji) [reply agent/agent_endpoint.go:1330]: I raised https://github.com/hashicorp/consul/issues/11393 for this issue! We prefer `return nil, err` over writing to response.

I'm not sure about the historic context, but in go's stdlib the handlers don't return any errors and you write to `resp` directly. This pattern may have been carried over to consul's handlers.

Comment (kisunji) [reply agent/agent_endpoint.go:1330]: Linking to @Mathew-Estafanous 's https://github.com/hashicorp/consul/pull/11599 as well who started work on the refactoring

Comment (jkirschner-hashicorp) [reply agent/agent_endpoint.go:1330]: @kisunji : thanks for creating that issue! If I understand correctly, `return nil, err` is what we want to do here and elsewhere. And applying elsewhere is captured in #11393. Yes? If so, do you want to do a final review/approve?

**Review by kisunji — APPROVED** [review]
LGTM!

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/539327.

**Comment by hc-github-team-consul-core** [comment]
:cherries::white_check_mark: Cherry pick of commit a36ddc31c7b9cac96eaf3c93ce9a724011fdbdc0 onto `release/1.11.x` succeeded!

**Pull request merged**
